src/reporters/Reporter.ts

```diff
@@ -0,0 +1,170 @@
+import 'istanbul'; /* import for side-effects */
+
+import { accessSync, constants, readFileSync } from 'fs';
+import { red, green, grey, white, bold } from 'chalk';
+import * as intern from 'intern';
+import * as Suite from 'intern/lib/Suite';
+import * as Test from 'intern/lib/Test';
+import * as Runner from 'intern/lib/reporters/Runner';
+import * as util from 'intern/lib/util';
+import * as Collector from 'istanbul/lib/collector';
+import * as JsonReporter from 'istanbul/lib/report/json';
+import * as nodeUtil from 'util';
+
+const DEFAULT_COVERAGE_FILENAME = 'coverage-final.json';
+
+module Reporter { }
+
+class Reporter extends Runner {
+	private _collector = new Collector();
+	private _errors: { [sessionId: string ]: any[] } = {};
+	private _filename: string;
+	private _output: NodeJS.WritableStream;
+	private _mode: string;
+
+	constructor(config: any = {}) {
+		super(config);
+
+		this._filename = config.file || DEFAULT_COVERAGE_FILENAME;
+
+		this.reporter = new JsonReporter({
+			file: this._filename,
+			watermarks: config.watermarks
+		});
+
+		this._output = config.output;
+
+		this._mode = config.mode || intern.mode;
+	}
+
+	private _writeCoverage(): void {
+
+		function checkCoverageFinal() {
+			try {
+				accessSync(DEFAULT_COVERAGE_FILENAME, constants.R_OK);
+			}
+			catch (e) {
+				return false;
+			}
+			return true;
+		}
+
+		if (checkCoverageFinal()) {
+			/* There is already coverage collected from a previous run */
+			this._collector.add(JSON.parse(readFileSync(this._filename, { encoding: 'utf8' })));
+		}
+
+		this.reporter.writeReport(this._collector, true);
+	}
+
+	coverage(sessionId: string, coverage: any) {
+		if (this._mode === 'client' || sessionId) {
+			const session = this.sessions[sessionId || ''];
+			session.coverage = true;
+			this._collector.add(coverage);
+		}
+	}
+
+	runEnd() {
+		let numEnvironments = 0;
+		let numTests = 0;
+		let numFailedTests = 0;
+		let numSkippedTests = 0;
+
+		for (const sessionId in this.sessions) {
+			const session = this.sessions[sessionId];
+			++numEnvironments;
+			numTests += session.suite.numTests;
+			numFailedTests += session.suite.numFailedTests;
+			numSkippedTests += session.suite.numSkippedTests;
+		}
+
+		console.log();
```

Comment:
`\n` can't be used elsewhere because the previous output is the tickmarks of the passing tests, which we don't know when they finish to append a newline.  So I have commented instead.